File: user-service/src/main/java/com/ilogos/user/common/TokenInfo.java
package com.ilogos.user.common;

import java.security.PublicKey;
import java.util.Date;
import java.util.UUID;

import org.springframework.security.oauth2.jwt.Jwt;

import com.ilogos.user.user.model.IUserBase;

import io.jsonwebtoken.Jwts;
import lombok.Getter;

@Getter
public class TokenInfo implements IUserBase {
    public final static String USERNAME_CLAIM = "username";
    public final static String EMAIL_CLAIM = "email";
    public final static String TYPE_CLAIM = "type";

    public final static String ACCESS_TYPE = "access";
    public final static String REFRESH_TYPE = "refresh";

    private final String token;
    private final Date expiration;
    private final String username;
    private final UUID id;
    private final String email;
    private final Date issuedAt;
    private final String type;

    public TokenInfo(String token, PublicKey publicKey) {
        this.token = token;
        var claims = Jwts.parserBuilder()
                .setSigningKey(publicKey)
                .build()
                .parseClaimsJws(token)
                .getBody();
        expiration = claims.getExpiration();
        username = claims.get(USERNAME_CLAIM, String.class);
        id = UUID.fromString(claims.getSubject());
        email = claims.get(EMAIL_CLAIM, String.class);
        issuedAt = claims.getIssuedAt();
        type = claims.get(TYPE_CLAIM, String.class);
    }

    public TokenInfo(Jwt jwt) {
        token = jwt.getTokenValue();

        expiration = Date.from(jwt.getExpiresAt());
        username = jwt.getClaimAsString(USERNAME_CLAIM);
        id = UUID.fromString(jwt.getSubject());
        email = jwt.getClaimAsString(EMAIL_CLAIM);
        issuedAt = Date.from(jwt.getIssuedAt());
        type = jwt.getClaimAsString(TYPE_CLAIM);
    }

    public boolean isExpired() {
        return expiration.before(new Date());
    }

    public boolean isValid(IUserBase user) {
        return user != null && user.getUsername().equals(username) && !isExpired();
    }

    public static boolean isAccessType(String type) {
        return ACCESS_TYPE.equals(type);
    }

    public boolean isAccessToken() {
        return isAccessType(type);
    }

    public static boolean isRefreshType(String type) {
        return REFRESH_TYPE.equals(type);
    }

    public boolean isRefreshToken() {
        return isRefreshType(type);
    }
}

File: course-service/src/main/kotlin/com/ilogos/course/utils/TokenInfo.kt
package com.ilogos.course.utils

import com.ilogos.course.user.IUser
import io.jsonwebtoken.Jwts
import org.springframework.security.oauth2.jwt.Jwt
import java.security.PublicKey
import java.util.*

class TokenInfo : IUser {

    companion object {
        const val USERNAME_CLAIM = "username"
        const val EMAIL_CLAIM = "email"
        const val TYPE_CLAIM = "type"

        const val ACCESS_TYPE = "access"
        const val REFRESH_TYPE = "refresh"

        fun isAccessType(type: String?): Boolean = type == ACCESS_TYPE
        fun isRefreshType(type: String?): Boolean = type == REFRESH_TYPE
    }

    val token: String
    val expiration: Date
    override val username: String
    override val id: UUID
    override val email: String
    val issuedAt: Date
    val type: String

    val isExpired: Boolean get() = expiration.before(Date())
    val isAccessToken: Boolean get() = isAccessType(type)
    val isRefreshToken: Boolean get() = isRefreshType(type)

    constructor(token: String, publicKey: PublicKey) {
        this.token = token
        val claims = Jwts.parserBuilder()
            .setSigningKey(publicKey)
            .build()
            .parseClaimsJws(token)
            .body

        expiration = claims.expiration
        username = claims[USERNAME_CLAIM, String::class.java]
        id = UUID.fromString(claims.subject)
        email = claims[EMAIL_CLAIM, String::class.java]
        issuedAt = claims.issuedAt
        type = claims[TYPE_CLAIM, String::class.java]
    }

    constructor(jwt: Jwt) {
        token = jwt.tokenValue
        expiration = Date.from(jwt.expiresAt)
        username = jwt.getClaimAsString(USERNAME_CLAIM) ?: ""
        id = UUID.fromString(jwt.subject)
        email = jwt.getClaimAsString(EMAIL_CLAIM) ?: ""
        issuedAt = Date.from(jwt.issuedAt)
        type = jwt.getClaimAsString(TYPE_CLAIM) ?: ""
    }

}
